mobile/src/main/java/org/openhab/habdroid/ui/WidgetAdapter.kt

```diff
@@ -812,38 +828,74 @@ class WidgetAdapter(
     }
 
     class ButtongridViewHolder internal constructor(private val initData: ViewHolderInitData) :
-        LabeledItemBaseViewHolder(initData, R.layout.widgetlist_buttongriditem), View.OnClickListener {
+        LabeledItemBaseViewHolder(initData, R.layout.widgetlist_buttongriditem),
+        View.OnClickListener,
+        View.OnTouchListener {
         private val table: GridLayout = itemView.findViewById(R.id.widget_content)
         private val spareViews = mutableListOf<MaterialButton>()
         private val maxColumns = itemView.resources.getInteger(R.integer.section_switch_max_buttons)
+        private var waitingForTouchRelease = false
+        private val buttonViews = mutableMapOf<String, MaterialButton>()
+
+        fun updateButton(button: Widget) {
+            buttonViews[button.id]?.let { buttonView ->
+                buttonView.setTextAndIcon(
+                    connection = connection, 
+                    label = button.label, 
+                    icon = button.icon,
+                    labelColor = button.labelColor,
+                    iconColor = button.iconColor,
+                    mapper = colorMapper
+                )
+                buttonView.isChecked = button.item?.state?.asString == button.command
+                boundWidget?.buttons?.replaceAll { if (it.id == button.id) button else it }
+            }
+        }
 
         override fun bind(widget: Widget) {
+            if (waitingForTouchRelease) {
```

Comment:
This is a problem, as the app will potentially show stale data in that case. Need to think about it...

Reply:
Depends on what "stale data" means. It handles button state updates through `ButtongridViewHolder.updateButton()`.

If however, during waiting, the whole buttongrid structure is changed (buttons added / removed), then possibly yes. I haven't tested this scenario but I don't think it's a normal operation.

Reply:
refactored so this is no longer done